File: rust/src/boxblur.rs
use num_cpus;
use std::sync::Arc;
use threadpool::ThreadPool;

fn index(x: usize, y: usize, side_size: usize) -> usize {
    x * side_size + y
}

fn index_to_2d(idx: usize, side_size: usize) -> (usize, usize) {
    (idx / side_size, idx % side_size)
}

fn generate_box_blur_kernel(size: usize) -> Vec<f64> {
    let num_of_elements = size * size;
    let mut kernel = vec![0.0; num_of_elements];

    for i in 0..num_of_elements {
        kernel[i] = 1.0 / (num_of_elements as f64);
    }

    kernel
}

fn apply_convolution(
    arr: &[u8],
    array_size: usize,
    element_idx: usize,
    kernel: &[f64],
    kernel_size: usize,
) -> u8 {
    let mut sum = 0.0;
    for i in 0..kernel_size {
        for j in 0..kernel_size {
            let (mut x, mut y) = index_to_2d(element_idx, array_size);
            x = x.wrapping_add(i).wrapping_sub(kernel_size / 2);
            y = y.wrapping_add(j).wrapping_sub(kernel_size / 2);
            if x >= array_size || y >= array_size {
                continue;
            }
            sum += f64::from(arr[index(x, y, array_size)]) * kernel[index(i, j, kernel_size)];
        }
    }
    sum.round() as u8
}

pub fn sequential(arr: Vec<u8>) -> Vec<u8> {
    const KERNEL_SIZE: usize = 19;
    let array_size = (arr.len() as f64).sqrt() as usize;

    let kernel = generate_box_blur_kernel(KERNEL_SIZE);

    (0..arr.len())
        .map(|idx| apply_convolution(&arr, array_size, idx, &kernel, KERNEL_SIZE))
        .collect()
}

pub fn parallel(arr: Vec<u8>) -> Vec<u8> {
    const KERNEL_SIZE: usize = 19;
    let array_size = (arr.len() as f64).sqrt() as usize;
    let num_elements = arr.len();

    let kernel = generate_box_blur_kernel(KERNEL_SIZE);

    // Determine the number of available CPUs
    let num_threads = num_cpus::get();
    let pool = ThreadPool::new(num_threads);

    // Shared result vector
    let result = Arc::new(arr.clone()); // Arc is holding the original array

    for idx in 0..num_elements {
        let kernel = kernel.clone();
        let result = Arc::clone(&result);
        pool.execute(move || {
            let value = apply_convolution(&result, array_size, idx, &kernel, KERNEL_SIZE);
            // Unsafe: Directly modify result without synchronization
            unsafe {
                let ptr = result.as_ptr().add(idx) as *mut u8;
                *ptr = value;
            }
        });
    }

    // Wait for all threads to finish
    pool.join();

    // Return the result
    Arc::try_unwrap(result).expect("Failed to unwrap Arc")
}

File: c/src/boxblur.c
#include "boxblur.h"

#define KERNEL_SIZE 19

size_t index(size_t x, size_t y, size_t side_size)
{
    return x * side_size + y;
}

void generate_box_blur_kernel(double *kernel, size_t size)
{
    double value = 1.0 / (size * size);
    for (size_t i = 0; i < size * size; ++i)
    {
        kernel[i] = value;
    }
}

unsigned char apply_convolution(const unsigned char *arr, size_t array_size, size_t element_idx, const double *kernel, size_t kernel_size)
{
    size_t kernel_half = kernel_size / 2;
    double sum = 0.0;

    size_t x = element_idx / array_size;
    size_t y = element_idx % array_size;

    for (size_t i = 0; i < kernel_size; i++)
    {
        for (size_t j = 0; j < kernel_size; j++)
        {
            size_t xi = x + i - kernel_half;
            size_t yj = y + j - kernel_half;
            if (xi < array_size && yj < array_size)
            {
                sum += arr[index(xi, yj, array_size)] * kernel[index(i, j, kernel_size)];
            }
        }
    }
    return (unsigned char)round(sum);
}

unsigned char *sequential(const unsigned char *arr, size_t arr_size)
{
    size_t array_size = (size_t)sqrt(arr_size);
    unsigned char *result = (unsigned char *)malloc(arr_size);
    if (!result)
    {
        return NULL;
    }

    double kernel[KERNEL_SIZE * KERNEL_SIZE];
    generate_box_blur_kernel(kernel, KERNEL_SIZE);

    for (size_t idx = 0; idx < arr_size; ++idx)
    {
        result[idx] = apply_convolution(arr, array_size, idx, kernel, KERNEL_SIZE);
    }

    return result;
}

unsigned char *parallel(const unsigned char *arr, size_t arr_size)
{
    size_t array_size = (size_t)sqrt(arr_size);
    unsigned char *result = (unsigned char *)malloc(arr_size);
    if (!result)
    {
        return NULL;
    }

    double kernel[KERNEL_SIZE * KERNEL_SIZE];
    generate_box_blur_kernel(kernel, KERNEL_SIZE);

#pragma omp parallel for
    for (size_t idx = 0; idx < arr_size; ++idx)
    {
        result[idx] = apply_convolution(arr, array_size, idx, kernel, KERNEL_SIZE);
    }

    return result;
}
